org.eclipse.jdt.ls.core/src/org/eclipse/jdt/ls/core/internal/commands/ProjectCommand.java

```diff
@@ -108,8 +110,10 @@
 	 *         modulepaths.
 	 * @throws CoreException
 	 * @throws URISyntaxException
+	 * @throws InterruptedException
 	 */
-	public static ClasspathResult getClasspaths(String uri, ClasspathOptions options) throws CoreException, URISyntaxException {
+	public static ClasspathResult getClasspaths(String uri, ClasspathOptions options) throws CoreException, URISyntaxException, InterruptedException {
+		JobHelpers.waitForJobsToComplete(new NullProgressMonitor());
```

Comment:
That's too broad. You'll wait for unrelated jobs to finish

Reply:
Can I narrow to:

```java
JobHelpers.waitForJobs(IConstants.UPDATE_PROJECT_FAMILY, new NullProgressMonitor());
JobHelpers.waitForBuildJobs(10 * 1000);
```

Also wait for build jobs since update request will also trigger a clean build in `MavenBuilder`. But return the result no matter build finishes or not after 10 secs.

Reply:
I don't think you need to wait for build jobs. During a build all the classpaths are supposed to have been set already, so building projects won't change anything to that

Reply:
Yes you are right. I found out that the case will fail after the invocation of `JavaCore.setClasspathContainer()`, which will trigger `SetContainerOperation`

I'm deciding to add following code:

```java
IWorkspace workspace = ResourcesPlugin.getWorkspace();
workspace.run(new IWorkspaceRunnable() {
    @Override
        public void run(IProgressMonitor monitor) {
    }
}, workspace.getRoot(), 0, new NullProgressMonitor());
```

to wait for it. (JavaModeChangeOperation will lock the workspace tree). Make sense?

Reply:
@snjeza your opinion on that?

Reply:
Just add more background here: The cases might fail with a very high possibility after [`BuildPathManager#mavenProjectChanged()`](https://github.com/eclipse/m2e-core/blob/master/org.eclipse.m2e.jdt/src/org/eclipse/m2e/jdt/internal/BuildPathManager.java#L188), where the `SetContainerOperation` will be triggered.

Reply:
you can try 
```
IWorkspace workspace = ResourcesPlugin.getWorkspace();
workspace.run(new IWorkspaceRunnable() {
    @Override
        public void run(IProgressMonitor monitor) {
    }
}, workspace.getRoot(), IWorkspace.AVOID_UPDATE, new NullProgressMonitor());
```